[clang][deps] Properly capture the global module and '\n' for all module directives

naveen-seth opened the pull request:
Previously, the newline after a module directive was not properly captured and printed by `clang::printDependencyDirectivesAsSource`.

According to [P1857R3](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1857r3.html), each directive must, after skipping horizontal whitespace, appear at the start of a logical line. Because the newline after module directives was missing, this invalidated the following line.
This fixes tests that were previously in violation of P1857R3, including Objective-C directives, which should also comply with P1857R3.

This also ensures that the global module fragment `module;` is captured by the dependency directives scanner.

naveen-seth committed:
[clang][deps] Recognize 'module;' in dependency directive scanner

With this change, the dependency directive scanner now properly identifies
"module;" as a directive, as per P1857R3.
Previously, the global module fragment was not recognized by the
scanner.

naveen-seth committed:
Add newline after 'module;' to be valid C++

naveen-seth committed:
[clang][deps] Properly capture the global module and '\n' for all module directives

Previously, the newline after a module directive was not properly
captured (or printed).

According to P1857R3, each directive must, after skipping horizontal
whitespace, be at the start of a logical line.
Because the newline after module directives was missing, this
invalidated the following line.

This also fixes or removes tests that were in violation of P1857R3.
This extends to Objective-C directives, which should also
follow P1857R3.

This also ensures that the global module fragment `module;` is
captured by the dependency directives scanner.

Bigcheese reviewed (approved):
Looks good to me, but I'd also like @akyrtzi to take a look if they have time.

akyrtzi commented on clang/unittests/Lex/DependencyDirectivesScannerTest.cpp:
```diff
@@ -639,15 +639,12 @@ TEST(MinimizeSourceToDependencyDirectivesTest, AtImport) {
   ASSERT_FALSE(minimizeSourceToDependencyDirectives(" @ import  A;\n", Out));
   EXPECT_STREQ("@import A;\n", Out.data());
 
-  ASSERT_FALSE(minimizeSourceToDependencyDirectives("@import A\n;", Out));
```
Why is this check removed?

naveen-seth replied:
From [P1857R3](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1857r3.html):

> The entire import or module directive (including the closing ;) must be on a single logical line and for module must not come from an #include.

While this paper addresses C++, I've consulted with @Bigcheese on whether Objective-C modules should follow the same rules.
In the same spirit, a line-splice was added to this test:
https://github.com/llvm/llvm-project/blob/586b4389f28ab98bcb1229144004b74e73cccf4b/clang/unittests/Lex/DependencyDirectivesScannerTest.cpp#L645-L647

This patch currently fails CI due to some `clang-scan-deps` test failures caused by these changes. I'll make sure to fix those issues now.
Sorry for missing this earlier.

naveen-seth committed:
For review only: additional `TryConsumeToken(tok::eod)`

Lines following the comment 'Note: Not required to pass CI' can be
removed without the CI failing.
The `TryConsumeToken(tok::eod)` is not required in those places, even
though we know that a tok::eod follows.
This is only for the patch review.

naveen-seth committed:
Address akyrtzi's review feedback

naveen-seth committed:
Update Parser and Preprocessor to handle the added tok::eod

This commit removes the extra `TryConsumeToken(tok::eod)` which were
added in e755d18c1cad and which are not explicitly required for this
fix or to pass the CI.

naveen-seth commented on clang/lib/Parse/Parser.cpp:
```diff
@@ -2361,6 +2361,8 @@ Parser::ParseModuleDecl(Sema::ModuleImportState &ImportState) {
   if (getLangOpts().CPlusPlusModules && Tok.is(tok::semi)) {
     SourceLocation SemiLoc = ConsumeToken();
+    // Note: Not required to pass CI
+    TryConsumeToken(tok::eod);
```
The patch passes the test-suite without the additional (commented) `TryConsumeToken(tok::eod)` calls added here, but since this patch adds a `tok::eod` at the end of every module directive, it also makes sense to consume it.

The next commit (https://github.com/llvm/llvm-project/pull/148685/commits/0bc7245a5f5543802d9e11510b21ed254ea8a2da) includes only the minimal changes and leaves out the unnecessary `TryConsumeToken(tok::eod)` calls.

I split this out into its own commit so I could get you feedback on this.

Bigcheese reviewed (approved):
This looks good to me, but I'd also like @yronglin to take a look since it interacts with their patch.

yronglin reviewed (approved):
LGTM! Thank you working on this.

naveen-seth commented:
Thank you for reviewing!
